quinn/src/lib.rs

```diff
@@ -798,93 +698,217 @@ impl Connection {
 
     /// The peer's UDP address.
     pub fn remote_address(&self) -> SocketAddr {
-        self.0
-            .endpoint
-            .borrow()
-            .inner
-            .connection(self.0.handle)
-            .remote()
-    }
-
-    /// The `ConnectionId`s defined for `conn` locally.
-    pub fn local_ids(&self) -> impl Iterator<Item = ConnectionId> {
-        self.0
-            .endpoint
-            .borrow()
-            .inner
-            .loc_cids(self.0.handle)
-            .cloned()
-            .collect::<Vec<_>>()
-            .into_iter()
+        self.0.borrow().inner.remote()
     }
+
     /// The `ConnectionId` defined for `conn` by the peer.
     pub fn remote_id(&self) -> ConnectionId {
-        self.0
-            .endpoint
-            .borrow()
-            .inner
-            .connection(self.0.handle)
-            .rem_cid()
+        self.0.borrow().inner.rem_cid()
     }
 
     /// The negotiated application protocol
     pub fn protocol(&self) -> Option<Box<[u8]>> {
-        self.0
-            .endpoint
-            .borrow()
-            .inner
-            .connection(self.0.handle)
-            .protocol()
-            .map(|x| x.into())
+        self.0.borrow().inner.protocol().map(|x| x.into())
     }
 
     // Update traffic keys spontaneously for testing purposes.
     #[doc(hidden)]
     pub fn force_key_update(&self) {
-        self.0
-            .endpoint
-            .borrow_mut()
-            .inner
-            .force_key_update(self.0.handle)
+        self.0.borrow_mut().inner.force_key_update()
     }
 }
 
 struct ConnectionInner {
-    endpoint: Rc<RefCell<EndpointInner>>,
+    log: Logger,
+    inner: quinn::Connection,
+    driver: Option<Task>,
     handle: ConnectionHandle,
+    pending: Pending,
     side: Side,
+    // TODO: Replace this with something custom that avoids using oneshots to cancel
+    timers: FuturesUnordered<Timer>,
```

Comment:
The debugging experience (and performance!) will probably be much improved if this is replaced with `[Option<tokio_timer::Delay>; quinn::Timer::COUNT]` directly; no need for a dynamic set now that it's per-connection.

Reply:
I'd like to defer that until after this commit/PR. 👍

Reply:
I was thinking it specifically might simplify diagnosing malfunctions blocking the merge, but I haven't looked closely at those yet.